src/lib/db.ts

```diff
@@ -80,4 +191,186 @@ export class DB {
     const userSubFolder = appMode === "authed" ? "" : `${this.firebaseUser ? this.firebaseUser.uid : "no-user-id"}/`;
     return `/${appMode}/${userSubFolder}`;
   }
+
+  //
+  // Paths
+  //
+
+  public getUserPath(user: UserModelType) {
+    return `users/${user.id}`;
+  }
+
+  public getClassPath(user: UserModelType) {
+    return `classes/${user.classHash}`;
+  }
+
+  public getOfferingPath(user: UserModelType) {
+    return `${this.getClassPath(user)}/offerings/${user.offeringId}`;
+  }
+
+  public getGroupsPath(user: UserModelType) {
+    return `${this.getOfferingPath(user)}/groups`;
+  }
+
+  public getGroupPath(user: UserModelType, groupId: string) {
+    return `${this.getGroupsPath(user)}/${groupId}`;
+  }
+
+  public getGroupUserPath(user: UserModelType, groupId: string, userId?: string) {
+    return `${this.getGroupPath(user, groupId)}/users/${userId || user.id}`;
+  }
+
+  //
+  // Listeners
+  //
+
+  private startListeners() {
+    return new Promise<void>((resolve, reject) => {
+      // listeners must start in this order so we know the latest group joined so we can autojoin groups if needed
+      this.startLatestGroupIdListener()
+        .then(() => {
+          return this.startGroupsListener();
+        })
+        .then(() => {
+          this.isListening = true;
+        })
+        .then(resolve)
+        .catch(reject);
+    });
+  }
+
+  private stopListeners() {
+    this.stopLatestGroupIdListener();
+    this.stopGroupListeners();
+    this.isListening = false;
+  }
+
+  private startLatestGroupIdListener() {
+    return new Promise<void>((resolve, reject) => {
+      const {user} = this.stores;
+      const latestGroupIdRef = this.latestGroupIdRef = this.getLatestGroupIdRef();
+      // use once() so we are ensured that latestGroupId is set before we resolve
+      latestGroupIdRef.once("value", (snapshot) => {
+        this.handleLatestGroupIdRef(snapshot);
+        latestGroupIdRef.on("value", this.handleLatestGroupIdRef);
+      })
+      .then(resolve)
```

Comment:
I got a types error on this line -- might be related to new versions of dependencies. I had to change it to:
```
.then(snapshot => {
  resolve();
})
```